### PR TITLE
Fix single precision cast to double in slic

### LINKED ISSUE #4274: [performance] Unwanted cast from float32 to double in slic
## Description

If a `float32` image is given to  `skimage.segmentation.slic`, it will be cast to `double` implicitly.

In fact, in `slic`, the Cython function `_slic_cython` is called with the input image preprocessed with `img_as_float`. If the input array data type is `float32`, it is not touched.
Since the `_slic_cython` signature is 
```
    (double[:, :, :, ::1] image_zyx,
     double[:, ::1] segments,
     float step,
     Py_ssize_t max_iter,
     double[::1] spacing,
     bint slic_zero)
```
the input image is automatically and silently cast to `double`.

Using the `np_floats` fused type as advised in #3111 solves the problem.

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

Closes #4274.

This PR the fused type `np_floats` in `_slic_cython` and care is taken in `slic` to not convert the input image to double precision if it is initially single precision.

Here are some measured timings showing the performance improvement:

```python
In [2]: from skimage.data import astronaut                                                               

In [3]: from skimage.util import img_as_float                                                            

In [4]: from skimage.segmentation import slic                                                            

In [5]: img = img_as_float(astronaut()[::2, ::2])                                                        

In [6]: img_f32 = img.astype('float32')                                                                  

In [7]: %timeit slic(img, n_segments=250, compactness=10, sigma=1)                                       
79 ms ± 248 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [8]: %timeit slic(img_f32, n_segments=250, compactness=10, sigma=1)                                   
74.6 ms ± 132 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: img = img_as_float(astronaut())                                                                  

In [10]: img_f32 = img.astype('float32')                                                                 

In [11]: %timeit slic(img, n_segments=250, compactness=10, sigma=1)                                      
309 ms ± 2.51 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [12]: %timeit slic(img_f32, n_segments=250, compactness=10, sigma=1)                                  
285 ms ± 973 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/segmentation/slic_superpixels.py [hunk 1]
```diff
@@ -141,20 +143,25 @@ def slic(image, n_segments=100, compactness=10., max_iter=10, sigma=0,
         if image.shape[-1] != 3 and convert2lab:
             raise ValueError("Lab colorspace conversion requires a RGB image.")
         elif image.shape[-1] == 3:
-            image = rgb2lab(image)
+            image = rgb2lab(image).astype(dtype)
```
Comment (rfezzani) [inline skimage/segmentation/slic_superpixels.py]: This conversion is useless once #4296 is merged

**Comment by emmanuelle** [comment]
Thanks @rfezzani and sorry it took so long to start a review. The speed gain is modest but it's good to limit the memory consumption. I was wondering whether one could add a test about the new change but I could not find something meaningful so :+1: from me. @scikit-image/core we need another review here.

**Comment by rfezzani** [comment]
Thank you @emmanuelle for your review.

**Comment by jni** [comment]
Thanks @rfezzani and apologies for the wait time! =)

**Comment by rfezzani** [comment]
Thank you @jni :-) better late then never  :laughing:

**Pull request merged**